cip/1.accepted/CIP2021-08-10-Operator-precedence.adoc

```diff
@@ -0,0 +1,727 @@
+- A case expression has multiple operands -- a value to test, for every case a test value/predicate and a result value, and a default value of the 'else' case.
+
+Some operands expect non-values, typically identifiers (variable names, function names, property keys, etc.).
+Such _non-value operands_ do not allow nesting of expressions because expression only result in values.
+
+However, most operands expect values of some types.
+Such _value operands_ allow nesting of all expressions that result in a values of the expected type.
+The nested expressions form an _operator tree_ and which typically evaluated bottom-up, i.e. all expressions that provide an operand are evaluated before the expression is evaluated that take this operands.
+(Higher-order expression, e.g. list comprehensions, have operands which are exceptions to this bottom-up evaluation.)
+In other words, the operator tree encodes the evaluation order.
+For instance, the following operator tree
+----
+    *
+   / \
+  +   4
+ / \
+3   2
+----
+encodes that the numeric addition is evaluated before the numeric multiplication, so that the expression results in `20`.
+
+However, the query string is a sequence of characters and not a tree.
+The parser turns the query string into such an operator tree.
+In the query string, nesting can clearly be denoted by parentheses.
+For instance, the query string
+----
+((3+2)*4)
+----
+allows a parser to unambiguously construct the operator tree shown above.
+
+For user convenience, better readability, and familiarity with common conventions in mathematical notation, openCypher (and most other query and programming languages) allows omitting the parentheses.
+For instance,
+----
+3+2*4
+----
+is also a valid expression.
+
+Without additional rules, however, it is not possible to unambiguously construct a operator tree for such an expression.
+In the particular case, two operator trees are possible:
+
+- `((3+2)*4)`
+- `(3+(2*4))`
+
+To avoid this kind of ambiguity, openCypher (and most other query and programming languages) has _precedence rules_.
+
+Precedence rules (or _rules of operator precedence_) define the order in which the different operators, i.e. the different kinds of expressions, are evaluated.
+For instance, numeric multiplication is evaluated before numeric addition, such that expression
+----
+3+2*4
+----
+is effectively evaluated as
+----
+(3+(2*4))
+----
+and, hence, results in `11` (rather than `20`).
+
+This CIP states the precedence rules of openCypher.
+
+== Proposal
+
+The precedence in openCypher is defined by (1) _precedence levels_ and (2) _operator associativity_.
+Both are defined in the following two subsections.
+
+=== Precedence levels
+
+The following <<precedenceLevels,table>> lists the precedence levels in descending order.
+
+.[[precedenceLevels]]Precedence levels
+[cols="<.<1a,<.<4a,<.<4a", options="header"]
+|===
+|Level         |Group         |Operators
+
+|12
```

Comment:
Depends who you ask (see, e.g. https://introcs.cs.princeton.edu/java/11precedence/ or https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence#table). It is likewise confusing to say the highest has the lowest number as it is confusing to have the highest level shown lowest in the table.

Probably for exactly this reason, many lists do not given any numbering at all. We can do the same here.

Reply:
Fair enough! I don't have a strong opinion, but I did find it convenient to have a level number to refer to when doing the reviewing. So I'd suggest keep it as is.

Reply:
As a minimal improvement, I have add a note before the table saying 

> The higher the level number the higher the precedence.

Hopefully that helps to avoid confusion.